spring-rabbit/src/main/java/org/springframework/amqp/rabbit/support/PublisherCallbackChannelImpl.java

```diff
@@ -691,32 +687,33 @@ private synchronized void generateNacksForPendingAcks(String cause) {
 				logger.debug("Added listener " + listener);
 			}
 		}
-		return this.pendingConfirms.get(listener);
 	}
 
 	@Override
-	public synchronized boolean removeListener(Listener listener) {
-		Listener mappedListener = this.listeners.remove(listener.getUUID());
-		boolean result = mappedListener != null;
-		if (result && this.listeners.size() == 0) {
-			this.delegate.removeConfirmListener(this);
-			this.delegate.removeReturnListener(this);
+	public synchronized Collection<PendingConfirm> expire(Listener listener, long cutoffTime) {
+		SortedMap<Long, PendingConfirm> pendingConfirmsForListener = this.pendingConfirms.get(listener);
+		if (pendingConfirmsForListener == null) {
+			return Collections.<PendingConfirm>emptyList();
 		}
-		Iterator<Entry<Long, Listener>> iterator = this.listenerForSeq.entrySet().iterator();
-		while (iterator.hasNext()) {
-			Entry<Long, Listener> entry = iterator.next();
-			if (entry.getValue() == listener) {
-				iterator.remove();
+		else {
+			synchronized (pendingConfirmsForListener) {
+				List<PendingConfirm> expired = new ArrayList<PendingConfirm>();
+				Iterator<Entry<Long, PendingConfirm>> iterator = pendingConfirmsForListener.entrySet().iterator();
+				while (iterator.hasNext()) {
+					PendingConfirm pendingConfirm = iterator.next().getValue();
+					if (pendingConfirm.getTimestamp() < cutoffTime) {
+						expired.add(pendingConfirm);
+						iterator.remove();
+					}
+					else {
+						break;
```

Comment:
Yes, it's a `SortedMap` keyed by the sequence number which is incremented by the rabbit client on each send.

See `RabbitTemplate.setupConfirm()`.

So, as soon as we hit a sequence that's not expired, the rest won't be either.

We have to keep them sorted because we might get a confirm that spans multiple sequence numbers (and listeners). See `processAck()`.